pulsar-client/src/test/java/org/apache/pulsar/client/impl/MessageIdAdvUtilsTest.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import static org.testng.Assert.assertEquals;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.BitSet;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.Cleanup;
+import org.apache.commons.lang3.RandomUtils;
+import org.apache.pulsar.client.api.MessageIdAdv;
+import org.testng.annotations.Test;
+
+/**
+ * Unit test for {@link MessageIdAdvUtils}.
+ */
+public class MessageIdAdvUtilsTest {
+
+    /**
+     * Call <code>acknowledge</code> concurrently with batch message, and verify that only return true once
+     *
+     * @see MessageIdAdvUtils#acknowledge(MessageIdAdv, boolean)
+     * @see MessageIdAdv#getAckSet()
+     */
+    @Test
+    public void testAcknowledgeIndividualConcurrently() throws InterruptedException {
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("pulsar-consumer-%d").build();
+        @Cleanup("shutdown")
+        ExecutorService executorService = Executors.newFixedThreadPool(16, threadFactory);
+        for (int i = 0; i < 100; i++) {
+            int batchSize = RandomUtils.nextInt(1, 32);
+            BitSet bitSet = new BitSet(batchSize);
+            bitSet.set(0, batchSize);
+            AtomicInteger individualAcked = new AtomicInteger();
+            CountDownLatch startLatch = new CountDownLatch(1);
+            CountDownLatch finishLatch = new CountDownLatch(batchSize);
+            for (int batchIndex = 0; batchIndex < batchSize; batchIndex++) {
+                BatchMessageIdImpl messageId = new BatchMessageIdImpl(1, 0, 0, batchIndex, batchSize, bitSet);
+                executorService.execute(() -> {
+                    try {
+                        startLatch.await();
```

Comment:
What is the purpose of `startLatch`? is it to maximize concurrency?
That doesn't seem to be achieved at the moment.

btw. There's a concurrency test here which might be a useful reference: https://github.com/apache/pulsar/blob/46c9ce9b068a0e0191a9598e6c863bda3f428bc2/pulsar-broker/src/test/java/org/apache/pulsar/broker/loadbalance/MultiBrokerLeaderElectionTest.java#L111-L141 
It uses java.util.concurrent.Phaser instead of CountDownLatch .

Reply:
The trigger of this issue is that messages within the same batch are acknowledged almost simultaneously. The `startLatch` is designed to align different user threads as closely as possible, thereby maximizing the likelihood of triggering a race condition.

Reply:
> There's a concurrency test here which might be a useful reference

I haven't used java.util.concurrent.Phaser before. I'll take some time to understand it. Thank you for reference.

Reply:
Changed to `java.util.concurrent.Phaser`, it's more suitable than `CountDownLatch` in [startingGate](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/util/concurrent/Phaser.html) use case.

Thanks @lhotari